insight/src/main/java/info/nightscout/androidaps/plugins/pump/insight/LocalInsightPlugin.java

```diff
@@ -1253,150 +1273,197 @@ private void processDateTimeChangedEvent(DateTimeChangedEvent event) {
+                false,
+                PumpSync.TemporaryBasalType.NORMAL,
+                event.getEventPosition(),
+                event.getEventPosition());
         temporaryBasals.add(temporaryBasal);
     }
 
     private void processEndOfTBREvent(String serial, List<TemporaryBasal> temporaryBasals, EndOfTBREvent event) {
         long timestamp = parseDate(event.getEventYear(), event.getEventMonth(), event.getEventDay(),
                 event.getEventHour(), event.getEventMinute(), event.getEventSecond()) + timeOffset;
-        InsightPumpID pumpID = new InsightPumpID();
-        pumpID.eventID = event.getEventPosition();
-        pumpID.pumpSerial = serial;
-        pumpID.eventType = "EndOfTBR";
-        pumpID.timestamp = timestamp;
-        databaseHelper.createOrUpdate(pumpID);
-        TemporaryBasal temporaryBasal = new TemporaryBasal(getInjector());
-        temporaryBasal.durationInMinutes = 0;
-        temporaryBasal.source = Source.PUMP;
-        temporaryBasal.pumpId = pumpID.id;
-        temporaryBasal.date = timestamp - 1500L;
+        insightDatabaseDao.createOrUpdate(new InsightPumpID(
+                timestamp - 1500L,
+                "EndOfTBR",
+                serial,
+                event.getEventPosition())
+        );
+
+        TemporaryBasal temporaryBasal = new PumpSync.PumpState.TemporaryBasal(
+                timestamp - 1500L,
+                0L,
+                100.0,
+                false,
+                PumpSync.TemporaryBasalType.NORMAL,
+                event.getEventPosition(),
+                event.getEventPosition());
         temporaryBasals.add(temporaryBasal);
     }
 
     private void processBolusProgrammedEvent(String serial, BolusProgrammedEvent event) {
         long timestamp = parseDate(event.getEventYear(), event.getEventMonth(), event.getEventDay(),
                 event.getEventHour(), event.getEventMinute(), event.getEventSecond()) + timeOffset;
-        InsightBolusID bolusID = databaseHelper.getInsightBolusID(serial, event.getBolusID(), timestamp);
-        if (bolusID != null && bolusID.endID != null) {
-            bolusID.startID = event.getEventPosition();
-            databaseHelper.createOrUpdate(bolusID);
+        InsightBolusID bolusID = insightDatabaseDao.getInsightBolusID(serial, event.getBolusID(), timestamp);
+        if (bolusID != null && bolusID.getEndID() != null) {                        // TODO() Check if test EndID is necessary
+            bolusID.setStartID(event.getEventPosition());
+            insightDatabaseDao.createOrUpdate(bolusID);
             return;
         }
-        if (bolusID == null || bolusID.startID != null) {
-            bolusID = new InsightBolusID();
-            bolusID.timestamp = timestamp;
-            bolusID.bolusID = event.getBolusID();
-            bolusID.pumpSerial = serial;
+        if (bolusID == null || bolusID.getStartID() != null) {                        // TODO() Check StartID test is necessary
```

Comment:
@TebbeUbben I don't understand here why you create a new bolusID if one is found with a StartID?
my first idea was to create a new bolusID in insight database only if `bolusID == null`

Reply:
In rare edge cases two boluses can share the same ID. We don't want to overwrite that.